#! /bin/sh /usr/share/dpatch/dpatch-run
## add-sqlite3-support.dpatch by  <diegolocal@localhost.localdomain>
##
## DP: Add sqlite3 support for freePBX. Note that most (if not) all this patch
## DP: is commited usptream, and should be part of freePBX 2.2.2 or 2.2.3

@DPATCH@
diff -urNad upstream~/amp_conf/bin/retrieve_conf upstream/amp_conf/bin/retrieve_conf
--- upstream~/amp_conf/bin/retrieve_conf	2007-05-08 13:34:28.000000000 +0300
+++ upstream/amp_conf/bin/retrieve_conf	2007-05-08 13:34:28.000000000 +0300
@@ -223,6 +223,18 @@
 		$db->connect( $DSN );
 		break;
 	
+	case "sqlite3":
+		if (!isset($amp_conf["AMPDBFILE"]))
+			die("You must setup properly AMPDBFILE in /etc/amportal.conf");
+			
+		if (isset($amp_conf["AMPDBFILE"]) == "")
+			die("AMPDBFILE in /etc/amportal.conf cannot be blank");
+
+		require_once('DB/sqlite3.php');
+		$datasource = "sqlite3:///" . $amp_conf["AMPDBFILE"] . "?mode=0666";
+		$db = DB::connect($datasource);
+		break;
+
 	default:
 		die( "Unknown SQL engine: [$db_engine]");
 }
diff -urNad upstream~/amp_conf/bin/retrieve_iax_conf_from_mysql.pl upstream/amp_conf/bin/retrieve_iax_conf_from_mysql.pl
--- upstream~/amp_conf/bin/retrieve_iax_conf_from_mysql.pl	2007-05-08 13:34:27.000000000 +0300
+++ upstream/amp_conf/bin/retrieve_iax_conf_from_mysql.pl	2007-05-08 13:34:28.000000000 +0300
@@ -60,6 +60,15 @@
 	my $db_file = $ampconf->{"AMPDBFILE"};
 	$dbh = DBI->connect("dbi:SQLite2:dbname=$db_file","","");
 }
+elsif ( $db_engine eq "sqlite3" ) {
+	if (!exists($ampconf->{"AMPDBFILE"})) {
+		print "No AMPDBFILE set in /etc/amportal.conf\n";
+		exit;
+	}
+	
+	my $db_file = $ampconf->{"AMPDBFILE"};
+	$dbh = DBI->connect("dbi:SQLite:dbname=$db_file","","");
+}
 
 # Load the 'register' lines into iax_registrations.conf
 
diff -urNad upstream~/amp_conf/bin/retrieve_op_conf_from_mysql.pl upstream/amp_conf/bin/retrieve_op_conf_from_mysql.pl
--- upstream~/amp_conf/bin/retrieve_op_conf_from_mysql.pl	2007-05-08 13:34:28.000000000 +0300
+++ upstream/amp_conf/bin/retrieve_op_conf_from_mysql.pl	2007-05-08 13:34:28.000000000 +0300
@@ -178,6 +178,15 @@
 	my $db_file = $ampconf->{"AMPDBFILE"};
 	$dbh = DBI->connect("dbi:SQLite2:dbname=$db_file","","");
 }
+elsif ( $db_engine eq "sqlite3" ) {
+	if (!exists($ampconf->{"AMPDBFILE"})) {
+		print "No AMPDBFILE set in /etc/amportal.conf\n";
+		exit;
+	}
+	
+	my $db_file = $ampconf->{"AMPDBFILE"};
+	$dbh = DBI->connect("dbi:SQLite:dbname=$db_file","","");
+}
 
 open( EXTEN, ">$op_conf" ) or die "Cannot create/overwrite config file: $op_conf ($!)\n";
 print EXTEN $warning_banner;
diff -urNad upstream~/amp_conf/bin/retrieve_queues_conf_from_mysql.pl upstream/amp_conf/bin/retrieve_queues_conf_from_mysql.pl
--- upstream~/amp_conf/bin/retrieve_queues_conf_from_mysql.pl	2007-05-08 13:34:27.000000000 +0300
+++ upstream/amp_conf/bin/retrieve_queues_conf_from_mysql.pl	2007-05-08 13:34:28.000000000 +0300
@@ -56,8 +56,18 @@
 	my $db_file = $ampconf->{"AMPDBFILE"};
 	$dbh = DBI->connect("dbi:SQLite2:dbname=$db_file","","");
 }
+elsif ( $db_engine eq "sqlite3" ) {
+	if (!exists($ampconf->{"AMPDBFILE"})) {
+		print "No AMPDBFILE set in /etc/amportal.conf\n";
+		exit;
+	}
+	
+	my $db_file = $ampconf->{"AMPDBFILE"};
+	$dbh = DBI->connect("dbi:SQLite:dbname=$db_file","","");
+}
+
+$statement = "SELECT keyword,data from $table_name where id=0 and keyword <> 'account'";
 
-$statement = "SELECT keyword,data from $table_name where id='-1' and keyword <> 'account'";
 my $result = $dbh->selectall_arrayref($statement);
 unless ($result) {
   # check for errors after every single database call
diff -urNad upstream~/amp_conf/bin/retrieve_sip_conf_from_mysql.pl upstream/amp_conf/bin/retrieve_sip_conf_from_mysql.pl
--- upstream~/amp_conf/bin/retrieve_sip_conf_from_mysql.pl	2007-05-08 13:34:27.000000000 +0300
+++ upstream/amp_conf/bin/retrieve_sip_conf_from_mysql.pl	2007-05-08 13:34:28.000000000 +0300
@@ -60,6 +60,15 @@
 	my $db_file = $ampconf->{"AMPDBFILE"};
 	$dbh = DBI->connect("dbi:SQLite2:dbname=$db_file","","");
 }
+elsif ( $db_engine eq "sqlite3" ) {
+	if (!exists($ampconf->{"AMPDBFILE"})) {
+		print "No AMPDBFILE set in /etc/amportal.conf\n";
+		exit;
+	}
+	
+	my $db_file = $ampconf->{"AMPDBFILE"};
+	$dbh = DBI->connect("dbi:SQLite:dbname=$db_file","","");
+}
 
 $statement = "SELECT keyword,data from $table_name where id=-1 and keyword <> 'account' and flags <> 1";
 my $result = $dbh->selectall_arrayref($statement);
diff -urNad upstream~/amp_conf/bin/retrieve_zap_conf_from_mysql.pl upstream/amp_conf/bin/retrieve_zap_conf_from_mysql.pl
--- upstream~/amp_conf/bin/retrieve_zap_conf_from_mysql.pl	2007-05-08 13:34:27.000000000 +0300
+++ upstream/amp_conf/bin/retrieve_zap_conf_from_mysql.pl	2007-05-08 13:34:28.000000000 +0300
@@ -61,6 +61,15 @@
 	my $db_file = $ampconf->{"AMPDBFILE"};
 	$dbh = DBI->connect("dbi:SQLite2:dbname=$db_file","","");
 }
+elsif ( $db_engine eq "sqlite3" ) {
+	if (!exists($ampconf->{"AMPDBFILE"})) {
+		print "No AMPDBFILE set in /etc/amportal.conf\n";
+		exit;
+	}
+	
+	my $db_file = $ampconf->{"AMPDBFILE"};
+	$dbh = DBI->connect("dbi:SQLite:dbname=$db_file","","");
+}
 
 $statement = "SELECT keyword,data from $table_name where id=-1 and keyword <> 'account' and flags <> 1";
 my $result = $dbh->selectall_arrayref($statement);
diff -urNad upstream~/amp_conf/htdocs/admin/common/db_connect.php upstream/amp_conf/htdocs/admin/common/db_connect.php
--- upstream~/amp_conf/htdocs/admin/common/db_connect.php	2006-06-25 13:37:15.000000000 +0300
+++ upstream/amp_conf/htdocs/admin/common/db_connect.php	2007-05-08 13:34:28.000000000 +0300
@@ -49,6 +49,18 @@
 		$db->connect( $DSN );
 		break;
 	
+	case "sqlite3":
+		if (!isset($amp_conf["AMPDBFILE"]))
+			die("You must setup properly AMPDBFILE in /etc/amportal.conf");
+			
+		if (isset($amp_conf["AMPDBFILE"]) == "")
+			die("AMPDBFILE in /etc/amportal.conf cannot be blank");
+
+		require_once('DB/sqlite3.php');
+		$datasource = "sqlite3:///" . $amp_conf["AMPDBFILE"] . "?mode=0666";
+		$db = DB::connect($datasource);
+		break;
+
 	default:
 		die( "Unknown SQL engine: [$db_engine]");
 }
diff -urNad upstream~/amp_conf/htdocs/admin/modules/core/functions.inc.php upstream/amp_conf/htdocs/admin/modules/core/functions.inc.php
--- upstream~/amp_conf/htdocs/admin/modules/core/functions.inc.php	2007-01-25 09:14:53.000000000 +0200
+++ upstream/amp_conf/htdocs/admin/modules/core/functions.inc.php	2007-05-08 13:38:04.000000000 +0300
@@ -1417,11 +1417,11 @@
 	global $db;
 	global $amp_conf;
 	
-	if ( $amp_conf["AMPDBENGINE"] == "sqlite")
+	if ( ($amp_conf["AMPDBENGINE"] == "sqlite") || ($amp_conf["AMPDBENGINE"] == "sqlite3") )
 	{
 		// TODO: sqlite work arround - diego 
 		// need to reorder the trunks in PHP code
-		$unique_trunks = sql("SELECT * FROM globals WHERE variable LIKE 'OUT_%' ORDER BY variable","getAll"); 
+                $unique_trunks = sql("SELECT * FROM globals WHERE variable LIKE 'OUT\\_%' escape '\\' ORDER BY variable","getAll");
 	}
 	else
 	{
@@ -1638,11 +1638,23 @@
 
 //get outbound routes for a given trunk
 function core_trunks_gettrunkroutes($trunknum) {
-	$results = sql("SELECT DISTINCT SUBSTRING(context,7), priority FROM extensions WHERE context LIKE 'outrt-%' AND (args LIKE 'dialout-trunk,".$trunknum.",%' OR args LIKE 'dialout-enum,".$trunknum.",%')ORDER BY context ","getAll");
+
+	global $amp_conf;
+
+	if ($amp_conf["AMPDBENGINE"] == "sqlite3")
+	        $results = sql("SELECT DISTINCT              context, priority FROM extensions WHERE context LIKE 'outrt-%' AND (args LIKE 'dialout-trunk,".$trunknum.",%' OR args LIKE 'dialout-enum,".$trunknum.",%')ORDER BY context ","getAll");
+	else	
+		$results = sql("SELECT DISTINCT SUBSTRING(context,7), priority FROM extensions WHERE context LIKE 'outrt-%' AND (args LIKE 'dialout-trunk,".$trunknum.",%' OR args LIKE 'dialout-enum,".$trunknum.",%')ORDER BY context ","getAll");
 	
 	foreach ($results as $row) {
-		$routes[$row[0]] = $row[1];
+                $t = ($amp_conf["AMPDBENGINE"] == "sqlite3") ? substr( $row[0], 7 ) : $row[0];
+                $r = $row[1];
+                $routes[ $r ] = $t;
+
+//		original code was:
+//		$routes[$row[0]] = $row[1];
 	}
+
 	// array(routename=>priority)
 	return isset($routes)?$routes:null;
 }
@@ -1666,10 +1678,16 @@
 {
 	global $amp_conf;
 	
-	if ( $amp_conf["AMPDBENGINE"] == "sqlite")
+	if ( ($amp_conf["AMPDBENGINE"] == "sqlite") || ($amp_conf["AMPDBENGINE"] == "sqlite3") )
 	{
-		// TODO: sqlite work arround - diego
+		// SUBSTRING is not supported under sqlite3, we need to filter
+		// this in php. I am not sure why "6" and not "7"
+		// but I don't really care -> it works :)
 		$results = sql("SELECT DISTINCT context FROM extensions WHERE context LIKE 'outrt-%' ORDER BY context ","getAll");
+		foreach( array_keys($results) as $idx )
+		{
+			$results[$idx][0] = substr( $results[$idx][0], 6);
+		}
 	}
 	else
 	{
@@ -1753,6 +1771,8 @@
 function core_routing_setroutepriority($routepriority, $reporoutedirection, $reporoutekey)
 {
 	global $db;
+	global $amp_conf;
+
 	$counter=-1;
 	foreach ($routepriority as $tresult) 
 	{
@@ -1812,13 +1832,19 @@
 			die($result->getMessage(). $sql); 
  		}
 	}
-	
-	$sql = "SELECT DISTINCT SUBSTRING(context,7) FROM extensions WHERE context LIKE 'outrt-%' ORDER BY context ";
+
+	if ( $amp_conf["AMPDBENGINE"] != "sqlite3")	
+		$sql = "SELECT DISTINCT SUBSTRING(context,7) FROM extensions WHERE context LIKE 'outrt-%' ORDER BY context ";
+	else
+		$sql = "SELECT DISTINCT context FROM extensions WHERE context LIKE 'outrt-%' ORDER BY context ";
+
         // we SUBSTRING() to remove "outrt-"
         $routepriority = $db->getAll($sql);
         if(DB::IsError($routepriority)) {
                 die($routepriority->getMessage());
         }
+
+	// TODO: strip the context on the sqlite3 backend
         return ($routepriority);
 	
 }
diff -urNad upstream~/install_amp upstream/install_amp
--- upstream~/install_amp	2007-05-08 13:34:28.000000000 +0300
+++ upstream/install_amp	2007-05-08 13:34:28.000000000 +0300
@@ -425,7 +425,8 @@
 				if ( (strtolower(substr($file,-4)) == ".sqlite") && ($db_engine == "sqlite") ) {
 					install_sqlupdate( $version, $file );
 				}
-				elseif ((strtolower(substr($file,-4)) == ".sql") && ( ($db_engine  == "mysql")  ||  ($db_engine  == "pgsql")) ) {
+				elseif ((strtolower(substr($file,-4)) == ".sql") && 
+						( ($db_engine  == "mysql")  ||  ($db_engine  == "pgsql") || ($db_engine == "sqlite3") ) ) {
 					install_sqlupdate( $version, $file );
 				}
 			}
@@ -983,6 +984,18 @@
 		$db->connect( $DSN );
 		break;
 	
+	case "sqlite3":
+		if (!isset($amp_conf["AMPDBFILE"]))
+			die("You must setup properly AMPDBFILE in /etc/amportal.conf");
+			
+		if (isset($amp_conf["AMPDBFILE"]) == "")
+			die("AMPDBFILE in /etc/amportal.conf cannot be blank");
+
+		require_once('DB/sqlite3.php');
+		$datasource = "sqlite3:///" . $amp_conf["AMPDBFILE"] . "?mode=0666";
+		$db = DB::connect($datasource);
+		break;
+
 	default:
 		die( "Unknown SQL engine: [$db_engine]");
 }
